src/TimeTrace/ExecutionHierarchy.cpp

```diff
@@ -0,0 +1,258 @@
+
+        return convertedString;
+    }
+
+    std::chrono::nanoseconds ConvertTime(long long ticks, long long frequency)
+    {
+        return std::chrono::nanoseconds{ Internal::ConvertTickPrecision(ticks, frequency, std::chrono::nanoseconds::period::den) };
+    }
+
+}  // anonymous namespace
+
+bool ExecutionHierarchy::Entry::OverlapsWith(const Entry* other) const
+{
+    return StartTimestamp < other->StopTimestamp &&
+           other->StartTimestamp < StopTimestamp;
+}
+
+ExecutionHierarchy::ExecutionHierarchy() :
+    entries_{},
+    roots_{},
+    symbolNames_{},
+    unresolvedTemplateInstantiationsPerSymbol_{}
+{
+}
+
+AnalysisControl ExecutionHierarchy::OnStartActivity(const EventStack& eventStack)
+{
+    if (   MatchEventStackInMemberFunction(eventStack, this, &ExecutionHierarchy::OnNestedActivity)
+        || MatchEventInMemberFunction(eventStack.Back(), this, &ExecutionHierarchy::OnRootActivity))
+    {}
+
+    if (   MatchEventInMemberFunction(eventStack.Back(), this, &ExecutionHierarchy::OnInvocation)
+        || MatchEventInMemberFunction(eventStack.Back(), this, &ExecutionHierarchy::OnFrontEndFile)
+        || MatchEventInMemberFunction(eventStack.Back(), this, &ExecutionHierarchy::OnFunction)
+        || MatchEventInMemberFunction(eventStack.Back(), this, &ExecutionHierarchy::OnTemplateInstantiation))
+    {}
+
+    return AnalysisControl::CONTINUE;
+}
+
+AnalysisControl ExecutionHierarchy::OnStopActivity(const EventStack& eventStack)
+{
+    if (MatchEventInMemberFunction(eventStack.Back(), this, &ExecutionHierarchy::OnFinishActivity))
+    {}
+
+    return AnalysisControl::CONTINUE;
+}
+
+AnalysisControl ExecutionHierarchy::OnSimpleEvent(const EventStack& eventStack)
+{
+    if (   MatchEventInMemberFunction(eventStack.Back(), this, &ExecutionHierarchy::OnSymbolName)
+        || MatchEventStackInMemberFunction(eventStack, this, &ExecutionHierarchy::OnCommandLine)
+        || MatchEventStackInMemberFunction(eventStack, this, &ExecutionHierarchy::OnEnvironmentVariable)
+        || MatchEventStackInMemberFunction(eventStack, this, &ExecutionHierarchy::OnFileInput)
+        || MatchEventStackInMemberFunction(eventStack, this, &ExecutionHierarchy::OnFileOutput))
+    {}
+
+    return AnalysisControl::CONTINUE;
+}
+
+const ExecutionHierarchy::Entry* ExecutionHierarchy::GetEntry(unsigned long long id) const
+{
+    auto it = entries_.find(id);
+    return it != entries_.end() ? &it->second : nullptr;
+}
+
+void ExecutionHierarchy::OnRootActivity(const Activity& root)
+{
+    Entry* entry = CreateEntry(root);
+    
+    assert(std::find(roots_.begin(), roots_.end(), entry) == roots_.end());
+    roots_.push_back(entry);
+}
+
+void ExecutionHierarchy::OnNestedActivity(const Activity& parent, const Activity& child)
+{
+    auto parentEntryIt = entries_.find(parent.EventInstanceId());
+    assert(parentEntryIt != entries_.end());
+
+    Entry* childEntry = CreateEntry(child);
```

Comment:
In reference to my Functions and TemplateInstantiations comment below, it would also be beneficial to not create a ton of entries for rejected functions and template instantiations.

Reply:
As mentioned in the comment for filtered `Function` activities, I'll move filtering to `ExecutionHierarchy` to save memory and don't process unnecessary entries.